tests/draft2019-09/recursiveRef.json

```diff
@@ -0,0 +1,310 @@
+                "valid": true
+            },
+            {
+                "description": "integer does not match as a property value",
+                "data": { "foo": 1 },
+                "valid": false
+            },
+            {
+                "description": "two levels, properties match with inner definition",
+                "data": { "foo": { "bar": "hi" } },
+                "valid": true
+            },
+            {
+                "description": "two levels, integer does not match as a property value",
+                "data": { "foo": { "bar": 1 } },
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "$recursiveRef with no $recursiveAnchor works like $ref",
+        "schema": {
+            "$id": "http://localhost:4242/recursiveRef5/schema.json",
+            "$defs": {
+                "myobject": {
+                    "$id": "myobject.json",
+                    "$recursiveAnchor": false,
+                    "anyOf": [
+                        { "type": "string" },
+                        {
+                            "type": "object",
+                            "additionalProperties": { "$recursiveRef": "#" }
+                        }
+                    ]
+                }
+            },
+            "anyOf": [
+                { "type": "integer" },
+                { "$ref": "#/$defs/myobject" }
+            ]
+        },
+        "tests": [
+            {
+                "description": "integer matches at the outer level",
+                "data": 1,
+                "valid": true
+            },
+            {
+                "description": "single level match",
+                "data": { "foo": "hi" },
+                "valid": true
+            },
+            {
+                "description": "integer does not match as a property value",
+                "data": { "foo": 1 },
+                "valid": false
+            },
+            {
+                "description": "two levels, properties match with inner definition",
+                "data": { "foo": { "bar": "hi" } },
+                "valid": true
+            },
+            {
+                "description": "two levels, integer does not match as a property value",
+                "data": { "foo": { "bar": 1 } },
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "$recursiveRef with no $recursiveAnchor in the initial target schema resource",
+        "schema": {
+            "$id": "http://localhost:4242/recursiveRef6/base.json",
+            "$recursiveAnchor": true,
+            "anyOf": [
+                { "type": "boolean" },
+                {
+                    "type": "object",
+                    "additionalProperties": {
+                        "$id": "http://localhost:4242/recursiveRef6/inner.json",
+                        "$comment": "there is no $recursiveAnchor: true here, so we do NOT recurse to the base",
```

Comment:
This `$comment` is wrong. It doesn't matter that there isn't a `$recursiveAnchor` in the embedded schema. The `$recursiveAnchor` that is set on line 239 still applies to this schema.

Reply:
How is the comment wrong?  The initial target of the $recursiveRef is here:  '#' resolves to inner.json. The rules for $recursiveRef explicitly says ""$recursiveRef" behaves identically to "$ref", except when its target schema contains "$recursiveAnchor" with a value of true."  So the initial target is the final target - we recurse to inner.json.

Reply:
@handrews, we were hoping you could help us resolve this. I previously commented on this PR the following.

> My understanding is that `$recursiveAnchor` is not intended to be a flag to enable or disable `$recursiveRef` functionality. It marks a schema as being allowed to be the target of a `$recursiveRef`. A `$recursiveRef` only behaves like `$ref` if there is no `$recursiveAnchor` anywhere in it's dynamic scope. Not every schema in the chain needs to have a `$recursiveAnchor`.

You indicated that that was correct.

However, after talking with @karenetheridge and re-reading the spec, it does sound like the spec says that a `$recursiveRef` can only consider the dynamic scope if there is a `$recursiveAnchor` in it's lexical scope.

If you could have a look at these tests and tell us how you expect them to validate, I think that should clear up what the intended behavior was.

Reply:
@jdesrosiers `$recursiveRef` has three steps IIRC

* evaluate as if `$ref` to find the preliminary target schema
* if the preliminary target schema has `$recursiveAnchor: true`, proceed to the next step, otherwise the preliminary target schema is the final target schema and the reference is considered resolved
* unwind the dynamic scope to find the outermost `$recursiveAnchor: true`, regardless of whether intermediate dynamic scopes also have `$recursiveAnchor: true`; this is the intermediate target schema
* find the base URI of the intermediate target schema, and re-resolve the value of `$recursiveRef` against that base URI - this produces the final target schema, and the reference is now resolved. _(although I think we limited the value of `$recursiveAnchor` to `#` or at least said it always had to end in an empty fragment, so in practice the final target schema is always the same as the intermediate target schema)_

Keep in mind that `$recursive*` have cases that don't work, notably as shown in `links.json` so there's no sensible complete test suite that both strictly conforms to the spec and works with `links.json`.  I forget the details but the keywords are at least a little broken, and I'm really hoping no one actually uses 2019-09 and had advised only providing a test suite for 2020-NN in order to encourage that.  But I guess we're supporting 2019-09 so 🤷 just be aware `$recursive*` doesn't entirely make sense.

Reply:
great, so given this, it sounds like all the test cases in this PR are correct as written (note I made a few amendments that will be squashed together in the merge commit). @jdesrosiers do you agree?

(Note: as I said on slack\#tests, I think it's perfectly acceptable to acknowledge that $recursiveRef isn't quite right in this draft, so one shouldn't try to do anything too complicated with it.)

Reply:
Thanks, @handrews. That's what I got from re-reading the spec.

> unwind the dynamic scope to find the outermost $recursiveAnchor: true, regardless of whether intermediate dynamic scopes also have $recursiveAnchor: true; this is the intermediate target schema

This is where our lines got crossed. We agreed on this part, but I was missing the part where we only do this step if we have `$recursiveAnchor` in the same schema as the `$recursiveRef`.

@karenetheridge these tests don't yet cover this case where we have an intermediate dynamic scope that doesn't have a `$recursiveAnchor`. It can be added later, but I wanted to point that out.

As far as I can tell, the two step resolve and `$recursiveAnchor` enabling/disabling `$recursiveRef` seem unnecessary. I can't tell why it doesn't just resolve the `$recursiveRef` against the current dynamic scope and be done. But, `$dynamicAnchor`/`$dynamicRef` handled this way better, so it doesn't matter any more.

Reply:
> but I was missing the part where we only do this step if we have `$recursiveAnchor` in the same schema as the `$recursiveRef`.

Correction: in the _initial target_ schema.  Not in the schema containing the `$recursiveRef` (which is the context or source or referring schema I don't remember what I called it)